Move connectivty tutorials (#2282)
giantswarm/docs: src/content/tutorials/connectivity/ingress/multi-nginx-ic/index.md

---
linkTitle: Multiple ingress nginx controllers
title: Running multiple ingress nginx controllers
description: Deploy multiple ingress nginx controllers in a Kubernetes cluster to separate different ingress traffic classes.
weight: 20
menu:
  principal:
    parent: tutorials-connectivity-ingress
    identifier: tutorials-connectivity-ingress-multi-ic
user_questions:
  - How do I install multiple ingress nginx controllers?
  - How do I separate internal and external Services?
  - How do I configure ingress nginx controller for internal traffic?
  - How do I override the NodePorts on KVM ingresses?
  - How do I configure ingress nginx controller to allow weak ciphers?
last_review_date: 2024-08-26
aliases:
  - /advanced/connectivity/ingress/multi-nginx-ic
  - /guides/multi-nginx/
  - /advanced/multi-nginx/
owner:
  - https://github.com/orgs/giantswarm/teams/team-cabbage
---

ingress nginx controller handles [ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) resources, routing traffic from outside the Kubernetes cluster to services within the cluster.

> Starting with [ingress nginx controller v1.8.0](/changes/managed-apps/nginx-ingress-controller-app/v1.8.0/), one can install multiple ingress nginx controllers in a Kubernetes cluster. The optional ingress nginx controller can be [installed as an App on your cluster]({{< relref "/getting-started/install-an-application#install-ingress-controller" >}}).

Reviewer: ```suggestion
It is possible to install multiple ingress controllers in a Kubernetes cluster. The ingress nginx controller can be [installed as an App on your cluster]({{< relref "/getting-started/install-an-application#install-ingress-controller" >}}).
```

> [ingress nginx controller v2.2.0](/changes/managed-apps/nginx-ingress-controller-app/v2.2.0/) will start installing an `IngressClass` with default name `nginx` and controller value `k8s.io/ingress-nginx`.

Reviewer: ```suggestion
ingress nginx installs an `IngressClass` with default name `nginx` and controller value `k8s.io/ingress-nginx`.
```

> It is possible to install multiple ingress controllers in a Kubernetes cluster. The ingress nginx controller can be [installed as an App on your cluster]({{< relref "/getting-started/install-an-application#install-ingress-controller" >}}).
> ingress nginx installs an `IngressClass` with default name `nginx` and controller value `k8s.io/ingress-nginx`.
> 
> Some use cases for this might be:
> 
> - An ingress controller that's behind an internal ELB for traffic between services within the VPC (or a group of peered VPCs)
> - An ingress controller behind an ELB that already terminates SSL
> - An ingress controller with different functionality or performance
> 

Author: ```suggestion
It's possible to install multiple ingress controllers in a Kubernetes cluster. The ingress nginx controller can be [installed as an App on your cluster]({{< relref "/getting-started/install-an-application#install-ingress-controller" >}}). Ingress nginx installs an `IngressClass` with the default name `nginx` and controller value `k8s.io/ingress-nginx`.

Some use cases for this might be:

- An ingress controller that's behind an internal ELB for traffic between services within the VPC (or a group of peered VPCs)
- An ingress controller behind an ELB that already terminates SSL
- An ingress controller with different functionality or performance

```

Most ingress nginx controller configuration options have controller-wide defaults. They can also be override on a per-ingress resource level.

In each case below, one installs a second ingress nginx controller with a different global-only configuration and a separate ingress class. Ingress resources managed by this ingress nginx controller installation can still be customized on a per-ingress resource level.

Further information on configuring ingress nginx controller can be found on the [Advanced ingress configuration]({{< relref "/tutorials/connectivity/ingress/configuration/index.md" >}}) page.

## Quick installation instructions for a second ingress nginx controller

1. Install a second ingress nginx controller App (and subsequent apps) with a different global-only configuration.
2. Change the `ingressClassName` to the appropriate ingress class name. Make sure the ingress class name and controller value of each ingress controller don't collide with each other.

## Set the ingress class name of each ingress

> __Note__ that if you are running multiple ingress controllers you need to use the appropriate `ingressClassName` in your ingress resources, for example.

Author: ```suggestion
__Note__: if you are running multiple ingress controllers, you need to use the appropriate `ingressClassName` in your ingress resources, for example.
```

This is how one can achieve it by using multiple ingress nginx controllers:

- Deploy one ingress nginx controller App using default settings, for making selected cluster services accessible via public internet
- Deploy second internal ingress nginx controller App. Use these user configuration overrides:

```yaml
controller:
  ingressClassResource:
    name: nginx-internal
    controllerValue: k8s.io/ingress-nginx-internal
  service:
    public: false
    subdomain: ingress-internal
    # Required for KVM only.
    # Don't set on AWS, Azure & others.
    nodePorts:
      http: 30012
      https: 30013
```

Each ingress nginx controller App installation has to have an unique ingress class. ingress resources can then declare which ingress controller should be handling their route definition by referencing the respective ingress class in the `ingressClassName` field. The default ingress nginx controller ingress class is `nginx`. In the above example we configure `nginx-internal` as the second ingress nginx controller installation's ingress class.

> On AWS and Azure, ingress nginx controller's `LoadBalancer` Service is fronted by the cloud provider's managed load balancer service. By default, ingress nginx controller will have a public load balancer. Changing `controller.service.public` flag to `false` declares that an internal load balancer should be created instead.

Author: ```suggestion
On AWS and Azure, ingress nginx controller's `LoadBalancer` service is fronted by cloud provider's managed load balancer service. By default, ingress nginx controller will have a public load balancer. Changing `controller.service.public` flag to `false` declares that an internal load balancer should be created instead.
```
